### PR TITLE
Add new option to preserveExternalChanges.

### PR DESCRIPTION
**Pull request by taeold**
This PR make some sweeping changes!

1) We change the default behavior of the SDK so that any un-annotated configuration option that is resettable are given `RESET_VALUE`. Practically, this means that given a function configuration like this:

```js
export const v2Req = onRequest(
  (req, res) => {
    res.sendStatus(200)
  },
);
```

And you go ahead and change the memory configuration of the function using GCP Console or `gcloud` tool, we will revert the memory back to platform default on the next function deploy.

2) You may optionally opt-out of this behavior by setting `preserveExternalChanges` option:

```js
exports.v1httpPreserve = functions
  .runWith({ preserveExternalChanges: true })
  .https.onRequest((req, resp) => {
    resp.status(200).send("PASS");
  });

functionsv2.setGlobalOptions({ preserveExternalChanges: true });

exports.v2http = functionsv2.https.onRequest((req, resp) => {
  resp.status(200).send("PASS");
});
```

This change broke a lot of test. I broke up the change for fixing test cases in a separate PR https://github.com/firebase/firebase-functions/pull/1254.

### CODE REVIEW

**Comment by taeold** [comment]
Current dependencies on/for this PR:
* master
  * **PR #1161** <a href="https://app.graphite.dev/github/pr/firebase/firebase-functions/1161" target="_blank">Graphite</a> 
    * **PR #1250** <a href="https://app.graphite.dev/github/pr/firebase/firebase-functions/1250" target="_blank">Graphite</a> 
      * **PR #1253** <a href="https://app.graphite.dev/github/pr/firebase/firebase-functions/1253" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/firebase/firebase-functions/1253?utm_source=stack-comment).

**Review by TheIronDev — APPROVED** [review]

File: src/common/options.ts [hunk 1]
```diff
@@ -40,3 +39,10 @@ export class ResetValue {
  * Special configuration value to reset configuration to platform default.
  */
 export const RESET_VALUE = ResetValue.getInstance();
+
+/**
+ * @internal
+ */
+export type ResettableKeys<T> = Required<{
+  [K in keyof T as [ResetValue] extends [T[K]] ? K : never]: null;
```
Comment (colerogers) [inline src/common/options.ts:47]: 😅

File: src/runtime/manifest.ts [hunk 2]
```diff
@@ -36,10 +36,12 @@ export interface ManifestEndpoint {
...
-  vpc?: {
-    connector: string | Expression<string> | ResetValue;
-    egressSettings?: string | Expression<string> | ResetValue;
-  };
+  vpc?:
+    | {
+        connector: string | Expression<string>;
+        egressSettings?: string | Expression<string> | ResetValue;
+      }
+    | ResetValue;
```
Comment (colerogers) [inline src/runtime/manifest.ts:44]: Is the reset of the whole vpc setting, just removing it?

**Review by colerogers — COMMENTED** [review]
Since `spec/v1/providers/fixtures.ts` is empty, can we remove it?

**Comment by taeold** [comment]
@colerogers It's used in the subsequent PR. Bad PR split!

File: src/common/options.ts (hunk 1 above)
Comment (taeold) [reply src/common/options.ts:47]: gotta love ts.

**Review by colerogers — APPROVED** [review]

**Pull request merged**